gneiss/balances.py

```diff
@@ -90,51 +102,63 @@ def balance_basis(tree_node):
 
 
 def _count_matrix(treenode):
-    n_tips = 0
-    nodes = list(treenode.levelorder(include_self=True))
-    # fill in the Ordered dictionary. Note that the
-    # elements of this Ordered dictionary are
-    # dictionaries.
-    counts = OrderedDict()
-    columns = ['k', 'r', 'l', 't', 'tips']
-    for n in nodes:
-        if n not in counts:
-            counts[n] = {}
-        for c in columns:
-            counts[n][c] = 0
-
-    # fill in r and l.  This is done in reverse level order.
-    for n in nodes[::-1]:
+    node_count = 0
+    for n in treenode.postorder(include_self=True):
+        node_count += 1
         if n.is_tip():
-            counts[n]['tips'] = 1
-            n_tips += 1
-        elif len(n.children) == 2:
-            lchild = n.children[0]
-            rchild = n.children[1]
-            counts[n]['r'] = counts[rchild]['tips']
-            counts[n]['l'] = counts[lchild]['tips']
-            counts[n]['tips'] = counts[n]['r'] + counts[n]['l']
+            n._tip_count = 1
         else:
-            raise ValueError("Not a strictly bifurcating tree!")
-
-    # fill in k and t
-    for n in nodes:
-        if n.parent is None:
-            counts[n]['k'] = 0
-            counts[n]['t'] = 0
-            continue
-        elif n.is_tip():
+            try:
+                left, right = n.children
+            except:
+                raise ValueError("Not a strictly bifurcating tree!")
+            n._tip_count = left._tip_count + right._tip_count
+
+    # TODO: use recarray
+    # col 0 -> right counts
+    # col 1 -> left counts
+    # col 2 -> k
```

Comment:
Given that example, it sounds like k is concerned with immediate
descendents and t is concerned with the descendents of its sibling? Or do
you mean to consider node 2?
On Jul 10, 2016 11:17 PM, "Jamie Morton" <notifications@github.com> wrote:

> In gneiss/balances.py
> <https://github.com/biocore/gneiss/pull/8#discussion_r70206318>:
>
> > -    for n in nodes:
> > -        if n.parent is None:
> > -            counts[n]['k'] = 0
> > -            counts[n]['t'] = 0
> > -            continue
> > -        elif n.is_tip():
> > +            try:
> > +                left, right = n.children
> > +            except:
> > +                raise ValueError("Not a strictly bifurcating tree!")
> > +            n._tip_count = left._tip_count + right._tip_count
> > +
> > +    # TODO: use recarray
> > +    # col 0 -> right counts
> > +    # col 1 -> left counts
> > +    # col 2 -> k
>
> To clarify, k is the number of tips left of the subtree.
>
> So if we have the following tree
>
>
>                                         /-a
>                               /5-------|
>                              |          \-b
>                     /3-------|
>                    |         |          /-f
>                    |          \6-------|
>           /2-------|                    \-d
>          |         |
>          |         |          /-h
> -1-------|          \4-------|
>          |                    \-j
>          |
>           \-l
>
> And we look at node 4
> k=1 since there is only 1 tip right of the subtree rooted at 4
> t=4 since there are 4 tips left of the subtree rooted at 4
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/biocore/gneiss/pull/8/files/23ad7ae6d93099e1a6575efac2e7500ac10833dd#r70206318>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AAc8slgTOpVspt_lBy_6WUtHJaHX48F2ks5qUeAPgaJpZM4JI_gR>
> .
>